common/src/api/internal/shared.rs

```diff
@@ -704,33 +705,142 @@ pub struct ResolvedVpcRouteSet {
 
 /// Describes the purpose of the dataset.
 #[derive(
-    Debug, Serialize, Deserialize, JsonSchema, Clone, Copy, PartialEq, Eq,
+    Debug, JsonSchema, Clone, PartialEq, Eq, Ord, PartialOrd, Hash, EnumCount,
 )]
-#[serde(rename_all = "snake_case")]
+#[serde(tag = "type", rename_all = "snake_case")]
 pub enum DatasetKind {
-    Crucible,
+    // Durable datasets for zones
+    #[serde(rename = "cockroachdb")]
     Cockroach,
+    Crucible,
     Clickhouse,
     ClickhouseKeeper,
     ExternalDns,
     InternalDns,
+
+    // Zone filesystems
+    ZoneRoot,
+    Zone {
+        name: String,
+    },
+
+    // Other datasets
+    Debug,
+}
+
+impl Serialize for DatasetKind {
```

Comment:
Oh hmm, I think you'll also want to update the `JsonSchema` impl to match this.

Reply:
Just to confirm -- what about the JsonSchema implementation should we update? It's currently being derived, and is used as part of internal APIs, but I don't think it *needs* to match the serialization + to/from string implementations.

Reply:
Updated the PR with a fixed JSON schema.

Basically the issue was that JSON schemas are meant to be a description of the serialization format, and they were out of sync since we'd manually implemented `Serialize` and `Deserialize`. It was currently moot for us because we were using `replace`, but it's a lurking bug that can lead to issues down the road.

I've changed the schema to be just `string`, which is overly generic but is at least correct. A more sophisticated impl would be something like `string` with a validation regex that lists all the possibilities, but that seems like overkill given that we're using `replace` anyway.